miner/miner.go

```diff
@@ -554,12 +556,67 @@ func (m *Miner) mineOne(ctx context.Context, base *MiningBase) (minedBlock *type
 	tProof := build.Clock.Now()
 
 	// get pending messages early,
-	msgs, err := m.api.MpoolSelect(context.TODO(), base.TipSet.Key(), ticket.Quality())
+	msgs, err := m.api.MpoolSelect(ctx, base.TipSet.Key(), ticket.Quality())
 	if err != nil {
 		err = xerrors.Errorf("failed to select messages for block: %w", err)
 		return nil, err
 	}
 
+	tEquivocateWait := build.Clock.Now()
+
+	// This next block exists to "catch" equivocating miners,
+	// who submit 2 blocks at the same height at different times in order to split the network.
+	// To safeguard against this, we make sure it's been EquivocationDelaySecs since our base was calculated,
+	// then re-calculate it.
+	// If the daemon detected equivocated blocks, those blocks will no longer be in the new base.
+	m.niceSleep(time.Until(base.ComputeTime.Add(time.Duration(build.EquivocationDelaySecs) * time.Second)))
+	newBase, err := m.GetBestMiningCandidate(ctx)
+	if err != nil {
+		err = xerrors.Errorf("failed to refresh best mining candidate: %w", err)
+		return nil, err
+	}
+
+	// If the base has changed, we take the _intersection_ of our old base and new base,
+	// thus ejecting blocks from any equivocating miners, without taking any new blocks.
+	if !newBase.TipSet.Equals(base.TipSet) {
+		newBaseMap := map[cid.Cid]struct{}{}
+		for _, newBaseBlk := range newBase.TipSet.Cids() {
+			newBaseMap[newBaseBlk] = struct{}{}
+		}
+
+		refreshedBaseBlocks := make([]*types.BlockHeader, 0, len(base.TipSet.Cids()))
+		for _, baseBlk := range base.TipSet.Blocks() {
+			if _, ok := newBaseMap[baseBlk.Cid()]; ok {
+				refreshedBaseBlocks = append(refreshedBaseBlocks, baseBlk)
+			}
+		}
+
+		if len(refreshedBaseBlocks) != len(base.TipSet.Blocks()) {
+			refreshedBase, err := types.NewTipSet(refreshedBaseBlocks)
+			if err != nil {
+				err = xerrors.Errorf("failed to create new tipset when refreshing: %w", err)
+				return nil, err
+			}
+
+			if !base.TipSet.MinTicket().Equals(refreshedBase.MinTicket()) {
+				ticket, err = m.computeTicket(ctx, &rbase, round, refreshedBase.MinTicket(), mbi)
```

Comment:
Isn't the ticket computed from drand and the power table at finality? If that's the case, how could the ticket ever change (assuming that the new base is always at the same round)

Reply:
We seem to also incorporate chain randomness (from the parent) into it: https://github.com/filecoin-project/lotus/blob/master/miner/miner.go#L599